controllers/suite_test.go

```diff
@@ -34,44 +41,100 @@ import (
+var _ = BeforeSuite(func(ctx SpecContext) {
 	logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
 
 	By("bootstrapping test environment")
+	timeout := 3 * time.Minute
 	testEnv = &envtest.Environment{
-		CRDDirectoryPaths:     []string{filepath.Join("..", "config", "crd", "bases")},
-		ErrorIfCRDPathMissing: true,
+		CRDDirectoryPaths:        []string{filepath.Join("..", "config", "crd", "bases")},
+		ErrorIfCRDPathMissing:    true,
+		ControlPlaneStartTimeout: timeout,
+		ControlPlaneStopTimeout:  timeout,
+		AttachControlPlaneOutput: false,
 	}
 
+	var cfg *rest.Config
 	var err error
-	// cfg is defined in this file globally.
-	cfg, err = testEnv.Start()
+	// this is a channel to signal when the test environment is ready
+	done := make(chan interface{})
+	go func() {
+		// this will block until the test environment is ready
+		defer GinkgoRecover()
+		cfg, err = testEnv.Start()
+		close(done)
+	}()
+	// wait for the test environment to be ready
+	Eventually(done).WithContext(ctx).WithTimeout(timeout).Should(BeClosed())
 	Expect(err).NotTo(HaveOccurred())
 	Expect(cfg).NotTo(BeNil())
 
-	err = corev1alpha1.AddToScheme(scheme.Scheme)
-	Expect(err).NotTo(HaveOccurred())
+	scheme := runtime.NewScheme()
+	Expect(corev1alpha1.AddToScheme(scheme)).To(Succeed())
+	Expect(k8sscheme.AddToScheme(scheme)).To(Succeed())
+	Expect(apiv1.AddToScheme(scheme)).To(Succeed())
 
 	//+kubebuilder:scaffold:scheme
 
-	k8sClient, err = client.New(cfg, client.Options{Scheme: scheme.Scheme})
+	k8sClient, err = client.New(cfg, client.Options{Scheme: scheme})
 	Expect(err).NotTo(HaveOccurred())
 	Expect(k8sClient).NotTo(BeNil())
 
+	By("Creating controller manager")
+	mgr, err = ctrl.NewManager(cfg, ctrl.Options{
+		Scheme:             scheme,
+		MetricsBindAddress: "0",
+		LeaderElection:     false,
+		Port:               8443,
+	})
+	Expect(err).ToNot(HaveOccurred())
+	Expect(mgr).ToNot(BeNil())
+
+	kc, err := kclient.NewClient("localhost:50051")
+	Expect(err).ToNot(HaveOccurred())
+
+	kcs := map[string]*kclient.Client{
+		"localhost:50051": kc,
+	}
+
+	By("Creatng the controllers")
+	k8sGPTController := &K8sGPTReconciler{
+		Client:        mgr.GetClient(),
+		Scheme:        mgr.GetScheme(),
+		K8sGPTClient:  kc,
+		k8sGPTClients: kcs,
+	}
+	Expect(k8sGPTController.SetupWithManager(mgr)).To(Succeed())
+
+	go func() {
+		defer GinkgoRecover()
+		ctrl.Log.Info("Starting the manager")
+		Expect(mgr.Start(ctrl.SetupSignalHandler())).To(Succeed())
+	}()
+
+	crd := &apiv1.CustomResourceDefinition{}
+	Expect(k8sClient.Get(ctx, types.NamespacedName{Name: "k8sgpts.core.k8sgpt.ai"}, crd)).To(Succeed())
+	Expect(k8sClient.Get(ctx, types.NamespacedName{Name: "results.core.k8sgpt.ai"}, crd)).To(Succeed())
+
 })
 
 var _ = AfterSuite(func() {
 	By("tearing down the test environment")
 	err := testEnv.Stop()
-	Expect(err).NotTo(HaveOccurred())
+	// Here we need to exit directly because the testEnv.Stop() may hang forever in some cases.
+	if err != nil {
```

Comment:
@AlexsJones Sometimes, testEnv will stop failing and I hit that issue before. See https://github.com/kubernetes-sigs/controller-runtime/issues/1571#issuecomment-1005575071 but I believe they have fixed this issue in this version. Is it hanging in your env?

Reply:
Mine is an M2 pro, so similar spec - shall we test to see if this hangs the GH action?

Reply:
I got your point, this PR also change the CI process to `make test`.  As you can see here are several times successful.

https://github.com/k8sgpt-ai/k8sgpt-operator/actions/runs/4976805323/jobs/8905206195?pr=99

```
KUBEBUILDER_ASSETS="/home/runner/work/k8sgpt-operator/k8sgpt-operator/bin/k8s/1.26.0-linux-amd64" go test ./... -coverprofile cover.out
?   	github.com/k8sgpt-ai/k8sgpt-operator	[no test files]
ok  	github.com/k8sgpt-ai/k8sgpt-operator/api/v1alpha1	0.037s	coverage: 58.0% of statements
?   	github.com/k8sgpt-ai/k8sgpt-operator/pkg/client	[no test files]
?   	github.com/k8sgpt-ai/k8sgpt-operator/pkg/common	[no test files]
?   	github.com/k8sgpt-ai/k8sgpt-operator/pkg/resources	[no test files]
?   	github.com/k8sgpt-ai/k8sgpt-operator/pkg/utils	[no test files]
ok  	github.com/k8sgpt-ai/k8sgpt-operator/controllers	66.830s	coverage: 25.2% of statements
```

Reply:
@AlexsJones Can we merge this one? I have tested this PR in my community's operator. It is without any hang.